lib/Sema/CSDiag.cpp

```diff
@@ -1064,16 +1128,24 @@ CalleeCandidateInfo::evaluateCloseness(Type candArgListType,
       // more parameters.
       // We can still do something more sophisticated with this.
       // FIXME: Use TC.isConvertibleTo?
-      if (rArgType->isEqual(paramType))
+
+      SmallVector<ArchetypeType *, 4> archetypes;
+      SmallVector<Type, 4> substitutions;
+      bool matched = findGenericSubstitutions(paramType, rArgType,
+                                              archetypes, substitutions);
+      
+      if (matched && archetypes.size() == 0)
         continue;
-      if (auto genericParam = paramType->getAs<GenericTypeParamType>())
-        paramType = genericParam->getDecl()->getArchetype();
-      if (paramType->is<ArchetypeType>() && !rArgType->hasTypeVariable()) {
+      if (matched && archetypes.size() == 1 && !rArgType->hasTypeVariable()) {
```

Comment:
Actually, I think you mean hasTypeParameter(). hasTypeVariable() is for if the type contains a TypeVariableType that gets inserted by the constraint system. By the time the solver is done there should be none left, but since this is in the couldn't-solve-it path, they can still exist.

This seems rare, but it does happen, like in test/Constraints/diagnostics.swift:644:
```
func foo() -> [Int] {
  return Array <Int> (count: 1) { // expected-error {{cannot invoke initializer for type 'Array<Int>' with an argument list of type '(count: Int, () -> _)'}}
    // expected-note @-1 {{expected an argument list of type '(count: Int, repeatedValue: Element)'}}
    return 1
  }
}
```
where rArgType (when checking the trailing closure param) is:
```
(function_type
  (input=tuple_type num_elements=0)
  (output=type_variable_type id=1))
```